src/main/java/fi/evident/raudikko/internal/morphology/WordParser.java

```diff
@@ -0,0 +1,232 @@
+        private @NotNull final StrongMorphemeWordPartBuilder currentStrongMorpheme = new StrongMorphemeWordPartBuilder();
+        private @NotNull final List<WordPart> wordParts = new ArrayList<>();
+
+        private void append(@NotNull CharSequence s) {
+            currentPart.append(s);
+        }
+
+        private @NotNull SingleWordPartBuilder getCurrentPart() {
+            return currentPart;
+        }
+
+        private void addBaseForm(@NotNull String baseForm) {
+            currentPart.addBaseForm(baseForm);
+        }
+
+        private void setAttribute(WordAttribute attribute) {
+            currentPart.addAttribute(attribute);
+        }
+
+        private void startWordPart() {
+            endWordPart();
+            currentPart.reset();
+        }
+
+        private void endWordPart() {
+            if (currentStrongMorpheme.isNotConsumed()) {
+                currentPart.addBaseForm(currentStrongMorpheme.popBaseForm());
+                currentStrongMorpheme.addPart(currentPart.toWordPart());
+
+                if (currentStrongMorpheme.isConsumed()) {
+                    wordParts.add(currentStrongMorpheme.toWordPart());
+                    currentStrongMorpheme.reset();
+                }
+            }
+
+            else if (currentPart.isNotEmpty())
+                wordParts.add(currentPart.toWordPart());
+        }
+
+        private void startStrongMorpheme(@NotNull String[] baseFormParts) {
+            currentStrongMorpheme.startStrongMorpheme(baseFormParts);
+        }
+
+        private @NotNull Word toWord() {
+            endWordPart();
+            return new Word(new ArrayList<>(wordParts));
+        }
+
+        private void reset() {
+            currentPart.reset();
+            wordParts.clear();
+            currentStrongMorpheme.reset();
+        }
+    }
+
+    private static final class StrongMorphemeWordPartBuilder {
+
+        private @NotNull String baseForm = "";
+        private @NotNull String[] baseFormParts = new String[0];
+        private final @NotNull List<SingleWordPart> parts = new ArrayList<>();
+
+        private void startStrongMorpheme(@NotNull String[] baseFormParts) {
+            this.baseForm = String.join("", baseFormParts);
+            this.baseFormParts = baseFormParts;
+        }
+
+        private @NotNull String popBaseForm() {
+            if (baseFormParts.length == 0)
+                throw new IllegalStateException("Base form parts empty");
+
+            String result = baseFormParts[0];
+            baseFormParts = Arrays.copyOfRange(baseFormParts, 1, baseFormParts.length);
+            return result;
+        }
+
+        private void addPart(@NotNull SingleWordPart part) {
+            parts.add(part);
+        }
+
+        private boolean isConsumed() {
+            return baseFormParts.length == 0;
+        }
+
+        private boolean isNotConsumed() {
+            return !isConsumed();
+        }
```

Comment:
Mä en tekisi omaa not-metodiaan vain yhden käyttökerran takia. Kyllä se `if (! currentStrongMorpheme.isConsumed())` olisi ihan tarpeeksi selkeä. Itse asiassa sanoisin että se on jopa selkeämpi, koska tässä koodissa tulee korostuu selkeämmin se että eka ja toinen `if`-testaavat täsmälleen saman asian arvoja.

```java
if (!currentStrongMorpheme.isConsumed()) {
    currentPart.addBaseForm(currentStrongMorpheme.popBaseForm());
    currentStrongMorpheme.addPart(currentPart.toWordPart());

    if (currentStrongMorpheme.isConsumed()) {
        wordParts.add(currentStrongMorpheme.toWordPart());
        currentStrongMorpheme.reset();
    }
}
```